Create coding tracker console app project

basemkasem opened the pull request.

basemkasem committed:
Create a new console project

basemkasem committed:
second test commit

basemkasem committed:
Create essential classes needed

basemkasem committed:
Created database and table
Created appconfig file to add connection string

basemkasem committed:
- Created main menu method.
- Created coding session model.
- Solved some wornings.

Commit:
Add coding tracker project

basemkasem committed:
Create README.md

Commit:
Delete unnecessery lines of code

Commit:
Merge branch 'main' of https://github.com/basemkasem/CodeReviews.Console.CodingTracker

Commit:
modufy readme file

Commit:
more of fixing readme file

basemkasem committed:
Merge branch 'main' into main

Commit:
Fix codacy issues

Dejmenek commented:
Hey @basemkasem 👋,
Thanks for submitting. I've reviewed your project and have added some comments.

___

Unfortunately, your project has the following issues marked as 🔴.
Suggestions are marked as 🟡.
Things I liked are marked as 🟢.
Please review them carefully to understand and resolve these issues, as they are related to the project requirements and are blocking my approval.

🔴 There is no validation to ensure date and time is in specific format.

![image](https://github.com/user-attachments/assets/8d9172dd-6f76-4904-90b9-7b789fdff693)

🔴 There is an issue with parsing the `StartTime` and `EndTime` fields from the database, causing retrieval errors for coding sessions.

![image](https://github.com/user-attachments/assets/b0692570-e420-4296-9db1-fc9eac2e45a4)

🔴 When there are no coding sessions in the database and the program asks to select coding session, pressing `Enter` causes an exception because there is nothing to select. It would be better to validate this by not prompting the user for a session when none are available, and inform them that they need to add a session first. The same applies when deleting a coding session.

![image](https://github.com/user-attachments/assets/e4acc155-f6d4-4d80-b012-28fd3d3cb007)
![image](https://github.com/user-attachments/assets/3536b47d-af7f-4fee-b728-25b1defbfe38)

___

🟡 UI/UX
- When there are no coding sessions, it would be nice to show something like "No coding sessions found" instead of displaying empty table with column names.

![image](https://github.com/user-attachments/assets/26fc9c89-368b-48c4-8c57-9733b787b5d1)

- When updating a coding session, it would be good to know what I'm updating. The user won't have to remember what was previously entered. You can show the record being updated.

🟡 Keep your date formats consistent - you're asking for input as `dd-MM-yyyy` or `hh:mm`, but in the database, coding sessions are saved and displayed in the `dd/MM/yyyy hh:mm:ss` format, pick one and stick to it.

🟡 Inside the while loop in `UpdateStartTime(int recordId)` and `UpdateEndTime(int recordId)`, you're calling `GetRecord(recordId)` , which queries the database to fetch the record. This means the query'll be executed every time the user provides an invalid input. You can avoid it by fetching the record once before entering the loop. This will avoid unnecessary queries and improve performance.

![image](https://github.com/user-attachments/assets/9f2d1ae2-90a4-460b-aef1-7ac389eebcac)
![image](https://github.com/user-attachments/assets/d383b4f9-c2b0-49aa-add4-a55122642500)

___

🟢 README file
Good job on writing this. It's perfect introduction to your app🤗

🟢 Parameterized Queries
![image](https://github.com/user-attachments/assets/78e7f9ad-cd66-49ea-be00-47bc87c7d097)

🟢 I like the use of enums as menu options

___
Feel free to reach out if you have any questions😁.

Commit:
-Fix validation issue
-Fix no sessions issues
-Add delete all sessions option
-Fix ui/ux issues
-Fix performance issue that read the record each time in the while loop.

Dejmenek commented:
🔴 The date format validation is functioning correctly. However, the program expects the user to enter the date in either the `dd-MM-yyyy hh:mm` or `hh:mm` format, while the ValidateTime function is checking for the `dd/MM/yyyy hh:mm` format😅

![image](https://github.com/user-attachments/assets/72bd9196-beca-4a15-9cd0-83b133886ef3)

🔴 I’m still encountering the same issue with parsing `StartTime` and `EndTime`😥. Could you please check if you’re seeing it on your side? 🤔
Here are the steps to reproduce the error:
- Add a new coding session with StartTime `23/02/2025 01:00` and EndTime `23/02/2025 02:00`
- Attempt to display all coding sessions

![image](https://github.com/user-attachments/assets/03e22791-9480-45b4-ab02-93e56ea8d1b7)

basemkasem commented:
- I fixed the first format problem.
- For the second one, it works fine with no exceptions
![image](https://github.com/user-attachments/assets/29cb39ad-e1f5-4b00-a062-d20f6ae395d4)

Commit:
Fix time format prompt

TheCSharpAcademy commented:
@Dejmenek Could you please try clearing the database and trying again? This is probably happening due to locale issues. SQLite will save the date differently depending on your system. It should work fine if all the dates are input by you. Let me know if it works?

TheCSharpAcademy commented:
@Dejmenek Approving this one and you'll get the points. 😄

@basemkasem @Radicals27 Project approved! 😄✅ It's a shame you didn't complete the challenges this time. 😟 It's always a good use of your time. But nothing stops you from revisiting the project later.

🥋And congratulations! You got your Olive Green Belt!! 🫒🫒

✋🏻Looking forward to seeing your next projects!
